crates/sncast/src/helpers/fee.rs

```diff
@@ -1,146 +1,183 @@
-                FeeSettings {
-                    max_gas: Some(NonZeroU64::try_from_(max_gas).map_err(anyhow::Error::msg)?),
-                    max_gas_unit_price: Some(
-                        NonZeroU128::try_from_(max_gas_unit_price).map_err(anyhow::Error::msg)?,
-                    ),
-                }
-            }
-            (Some(max_fee), Some(max_gas), None) => {
-                if max_fee < max_gas {
-                    bail!("--max-fee should be greater than or equal to --max-gas amount");
-                }
-
-                let max_gas_unit_price = NonZeroFelt::try_from(Felt::from(max_fee).floor_div(&max_gas))
-                        .unwrap_or_else(|_| unreachable!("Calculated max gas unit price must be >= 1 because max_fee >= max_gas ensures a positive result"));
-                print_max_fee_conversion_info(max_fee, max_gas, max_gas_unit_price);
-                FeeSettings {
-                    max_gas: Some(NonZeroU64::try_from_(max_gas).map_err(anyhow::Error::msg)?),
-                    max_gas_unit_price: Some(
-                        NonZeroU128::try_from_(max_gas_unit_price).map_err(anyhow::Error::msg)?,
-                    ),
-                }
-            }
-            (Some(max_fee), None, None) => {
-                let max_gas_unit_price = NonZeroFelt::try_from(
-                    provider
-                        .get_block_with_tx_hashes(block_id)
-                        .await?
-                        .l1_gas_price()
-                        .price_in_fri,
-                )?;
-                // TODO(#2852)
-                let max_gas = NonZeroFelt::try_from(Felt::from(max_fee)
-                            .floor_div(&max_gas_unit_price)).context("Calculated max-gas from provided --max-fee and the current network gas price is 0. Please increase --max-fee to obtain a positive gas amount")?;
-                print_max_fee_conversion_info(max_fee, max_gas, max_gas_unit_price);
-                FeeSettings {
-                    max_gas: Some(NonZeroU64::try_from_(max_gas).map_err(anyhow::Error::msg)?),
-                    max_gas_unit_price: Some(
-                        NonZeroU128::try_from_(max_gas_unit_price).map_err(anyhow::Error::msg)?,
-                    ),
-                }
-            }
-        };
+    pub fn try_into_fee_settings(&self, fee_estimate: Option<&FeeEstimate>) -> Result<FeeSettings> {
+        // If some resource bounds values are lacking, starknet-rs will estimate them automatically
+        // but in case someone passes --max-fee flag, we need to make estimation on our own
+        // to check if the fee estimate isn't higher than provided max fee
+        if let Some(max_fee) = self.max_fee {
+            self.validate_max_fee_meets_resource_bounds()?;
+
+            let fee_estimate = fee_estimate
+                .ok_or_else(|| anyhow!("Fee estimate must be passed when max_fee is provided"))?;
```

Comment:
Actually, now that I look at it, it doesn't make sense to bubble it as a handleable error. If this error handling code is reached it means we have a bug in our logic somewhere else and this should never be allowed. Let's replace it with `expect()`.